File: packages/java/src/main/java/com/tesote/sdk/models/TransactionOrder.java
package com.tesote.sdk.models;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.math.BigDecimal;
import java.util.Map;

/**
 * Transaction order — the unit of payment intent in v2 batches and
 * transaction-order endpoints.
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public record TransactionOrder(
        @JsonProperty("id") String id,
        @JsonProperty("status") String status,
        @JsonProperty("amount") BigDecimal amount,
        @JsonProperty("currency") String currency,
        @JsonProperty("description") String description,
        @JsonProperty("reference") String reference,
        @JsonProperty("external_reference") String externalReference,
        @JsonProperty("idempotency_key") String idempotencyKey,
        @JsonProperty("batch_id") String batchId,
        @JsonProperty("scheduled_for") String scheduledFor,
        @JsonProperty("approved_at") String approvedAt,
        @JsonProperty("submitted_at") String submittedAt,
        @JsonProperty("completed_at") String completedAt,
        @JsonProperty("failed_at") String failedAt,
        @JsonProperty("cancelled_at") String cancelledAt,
        @JsonProperty("source_account") SourceAccount sourceAccount,
        @JsonProperty("destination") Destination destination,
        @JsonProperty("fee") Fee fee,
        @JsonProperty("execution_strategy") String executionStrategy,
        @JsonProperty("tesote_transaction") TesoteTransaction tesoteTransaction,
        @JsonProperty("latest_attempt") LatestAttempt latestAttempt,
        @JsonProperty("metadata") Map<String, Object> metadata,
        @JsonProperty("created_at") String createdAt,
        @JsonProperty("updated_at") String updatedAt
) {
    @JsonIgnoreProperties(ignoreUnknown = true)
    public record SourceAccount(
            @JsonProperty("id") String id,
            @JsonProperty("name") String name,
            @JsonProperty("payment_method_id") String paymentMethodId
    ) {}

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record Destination(
            @JsonProperty("payment_method_id") String paymentMethodId,
            @JsonProperty("counterparty_id") String counterpartyId,
            @JsonProperty("counterparty_name") String counterpartyName
    ) {}

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record Fee(
            @JsonProperty("amount") BigDecimal amount,
            @JsonProperty("currency") String currency
    ) {}

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record TesoteTransaction(
            @JsonProperty("id") String id,
            @JsonProperty("status") String status
    ) {}

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record LatestAttempt(
            @JsonProperty("id") String id,
            @JsonProperty("status") String status,
            @JsonProperty("attempt_number") Integer attemptNumber,
            @JsonProperty("external_reference") String externalReference,
            @JsonProperty("submitted_at") String submittedAt,
            @JsonProperty("completed_at") String completedAt,
            @JsonProperty("error_code") String errorCode,
            @JsonProperty("error_message") String errorMessage
    ) {}
}

File: packages/ts/src/models/transaction_order.ts
/**
 * TransactionOrder — outbound payment lifecycle object.
 * State machine: draft -> pending_approval -> approved -> processing -> completed|failed|cancelled.
 */

export type TransactionOrderStatus =
  | 'draft'
  | 'pending_approval'
  | 'approved'
  | 'processing'
  | 'completed'
  | 'failed'
  | 'cancelled';

export interface Beneficiary {
  name: string;
  bank_code?: string | null;
  account_number?: string | null;
  identification_type?: string | null;
  identification_number?: string | null;
}

export interface TransactionOrderSourceAccount {
  id: string;
  name: string;
  payment_method_id: string;
}

export interface TransactionOrderDestination {
  payment_method_id: string;
  counterparty_id: string;
  counterparty_name: string;
}

export interface TransactionOrderFee {
  amount: number;
  currency: string;
}

export interface TransactionOrderTesoteTransaction {
  id: string;
  status: string;
}

export interface TransactionOrderLatestAttempt {
  id: string;
  status: string;
  attempt_number: number;
  external_reference: string | null;
  submitted_at: string | null;
  completed_at: string | null;
  error_code: string | null;
  error_message: string | null;
}

export interface TransactionOrder {
  id: string;
  status: TransactionOrderStatus;
  amount: number;
  currency: string;
  description: string;
  reference: string | null;
  external_reference: string | null;
  idempotency_key: string | null;
  batch_id: string | null;
  scheduled_for: string | null;
  approved_at: string | null;
  submitted_at: string | null;
  completed_at: string | null;
  failed_at: string | null;
  cancelled_at: string | null;
  source_account: TransactionOrderSourceAccount;
  destination: TransactionOrderDestination;
  /** Null when fee_cents is zero. */
  fee: TransactionOrderFee | null;
  execution_strategy: string | null;
  tesote_transaction: TransactionOrderTesoteTransaction | null;
  latest_attempt: TransactionOrderLatestAttempt | null;
  created_at: string;
  updated_at: string;
}

export interface TransactionOrderListResponse {
  items: TransactionOrder[];
  has_more: boolean;
  limit: number;
  offset: number;
}
